core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1364,7 +1370,9 @@ private Engine createNewEngine(EngineConfig config) {
     }
 
     protected Engine newEngine(EngineConfig config) {
-        return engineFactory.newReadWriteEngine(config);
+        Engine engine = engineFactory.newReadWriteEngine(config);
+        refreshListeners.listenTo(engine);
```

Comment:
I also think the relationship is inverted, engine config should just accept a listener